bip-hugonguyen-bsms.mediawiki

```diff
@@ -0,0 +1,394 @@
+* The Signer generates a key record by prompting the user for a multisig derivation path and retrieves the XPUB at that derivation path. Alternatively, the Signer can choose a path on behalf of the user. If the Signer chooses the path, it should try to avoid reusing XPUBs for different wallets.
+* The first line in the record must be the specification version (<tt>BSMS 1.0</tt> as of this writing). The second line must be the hex-encoded <tt>TOKEN</tt>. The third line must be the <tt>KEY</tt>. The <tt>KEY</tt> is an XPUB plus its key origin information, written in the descriptor-defined format, i.e.: <tt>[{master key fingerprint}/{derivation path}]{XPUB}</tt>. The fourth line is a text description of the key, 80 characters maximum. The fifth line must be a <tt>SIG</tt>, whereas <tt>SIG</tt> is the signature generated by using the private key associated with the XPUB to sign the first four lines. The signature should follow [https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki BIP-0322], legacy format accepted.
+* The Signer calculates the Message Authentication Code (<tt>MAC</tt>) for the record. The first 16 bytes of the <tt>MAC</tt> serves as the Initialization Vector (<tt>IV</tt>) for the encryption.
+* The Signer encrypts the key record with the <tt>ENCRYPTION_KEY</tt> and <tt>IV</tt>.
+* The Signer encodes the <tt>MAC</tt> and the ciphertext into hexadecimal format, then concatenates the results: <tt>(MAC || ciphertext)</tt>.
+
+====Round 2====
+
+=====Coordinator=====
+
+* The Coordinator gathers key records from all participating Signers. The Coordinator verifies that there are exactly <tt>N</tt> unique key records before the wallet setup session expires.
+* For each key record, the Coordinator extracts the <tt>MAC</tt> from the data, sets <tt>IV</tt> to the first 16 bytes of the <tt>MAC</tt>, then decrypts the ciphertext using the <tt>ENCRYPTION_KEY</tt> and <tt>IV</tt>.
+* The Coordinator verifies that the included <tt>MAC</tt> is valid given the plaintext.
+* The Coordinator verifies that the key records have compatible specification versions.
+* The Coordinator verifies that the included <tt>SIG</tt> is valid given the <tt>KEY</tt>.
+* If all key records look good, the Coordinator fills in all necessary information to generate a descriptor record.
+* The first line in the descriptor record must be the specification version (<tt>BSMS 1.0</tt> as of this writing). The second line must be a comma-separated list of accepted derivation paths that the Signers can use to generate addresses from the included XPUBs. The paths must start with <tt>/</tt> and use non-hardened derivation. For example, <tt>/0/*</tt> and <tt>/1/*</tt> are some common paths. If there are no restrictions, it must say <tt>No path restrictions</tt>. The third line must be the descriptor string plus a <tt>CHECKSUM</tt>, all in one line. The <tt>CHECKSUM</tt> has [https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#checksums BECH32 encoding].
```

Comment:
A minor note: while the third line is _syntactically_ a valid descriptor as per Bitcoin Core, it isn't _semantically_. For example, `wsh(sortedmulti(2,keyA/*,keyB/*))#checksum` describes the list `[wsh(sortedmulti(2,keyA/0,keyB/0)), wsh(sortedmulti(2,keyA/1,keyB/1))], wsh(sortedmulti(2,keyA/2,keyB/2)), ...]`, which is very different from what it represents in the context of BSMS. Therefore, calling that a _descriptor_ might generate confusion, as [it has already happened](https://github.com/nunchuk-io/bips/pull/1/#issuecomment-804152141) in the discussion.

The idea of separating the valid derivation paths from the output descriptors is great, as it works well for any multisig scenario, and as far as I understand it seamlessly generalizes to miniscript policies).
 
Therefore, I think it might be worth finding a new term (for example _descriptor template_; or _wallet descriptor_) and have a short paragraph in the introduction to explain how the _descriptor template_ is expanded with the possible derivation paths to obtain the full list of output script descriptors, which would harmonize the naming with Core.
Such descriptor template, as it currently stands, would be exactly like a descriptor, except it redefines the meaning of the `*` symbol.

Reply:
+1 on descriptor template or similar naming.  So many vendors/applications do this, so it would be good to try standardize this and not confuse with the same naming

Reply:
I don't have a strong opinion on this, since it's just a very small difference in terminology. Calling that line a "descriptor template" is fine with me. But IMHO calling it just a "descriptor" is not wrong either, syntactically and semantically.

> Such descriptor template, as it currently stands, would be exactly like a descriptor, except it redefines the meaning of the `*` symbol.

Not quite. We are not redefining the meaning of `/*`. `/*` simply means that the Signers are entering a contract that accepts all non-hardened keys/addresses under the XPUBs being submitted.

An address generated from `/0/*` belongs to the address space of `/*` as well.

One can think of `/0/*` and `/1/*` as conventions that parties agree among themselves to reduce the usable address space within `/*`.

Reply:
> An address generated from `/0/*` belongs to the address space of `/*` as well.
> 
> One can think of `/0/*` and `/1/*` as conventions that parties agree among themselves to reduce the usable address space within `/*`.

This is _not_ what `*` means in output descriptors, and the precise reason you want a different term (if not a different symbol altogether).

In hindsight, I would have preferred to have the `#` symbol in output descriptors for the _address index_, and reserve `*` for use cases like yours.

Reply:
Another option could be to use `\**` for the descriptor template, which is a standard way of matching _all subpaths_ when describing paths in a filesystem (e.g. in `.gitignore`), which is a familiar semantics for many people. Moreover, it would be visually different and an invalid descriptor. This is desirable in my opinion, since entering the descriptor template into a descriptor wallet would _not_ find any valid addresses!

This might also address the concerns that @Rspigler and @prusnak voiced multiple times.

Reply:
> > An address generated from `/0/*` belongs to the address space of `/*` as well.
> > One can think of `/0/*` and `/1/*` as conventions that parties agree among themselves to reduce the usable address space within `/*`.
> 
> This is _not_ what `*` means in output descriptors

What do you mean? I'm interpreting `*` here as per defined in https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md:

`xpub encoded extended public key or xprv encoded extended private key (as defined in BIP 32).`
`Optionally followed by a single /* or /*' final step **to denote all (direct) unhardened or hardened children.**`

`/*` literally means all unhardened children, of which `/0/*` and `/1/*` are subsets.

Reply:
> Another option could be to use `\**` for the descriptor template, which is a standard way of matching _all subpaths_ when describing paths in a filesystem (e.g. in `.gitignore`), which is a familiar semantics for many people.

That is not standard as per the current descriptor language.

Reply:
_direct_, as in _exactly 1 level of derivation_. Perhaps that page should be clarified as well, I had the same confusion the first time I read it.

The rationale is: a descriptor wallet needs to give enough information to _easily_ find the described addresses. Therefore, you want a 1-dimensional list you can iterate.

Reply:
> _direct_, as in _exactly 1 level of derivation_. Perhaps that page should be clarified as well, I had the same confusion the first time I read it.

I actually missed this. Good point, I stand corrected.

Reply:
In that case, I agree we should rename this a "descriptor template", and have a short explanation of what it is.

Reply:
I like this, and will update https://github.com/bitcoin/bips/pull/1089 to use the same terminology

Reply:
done https://github.com/bitcoin/bips/pull/1097/commits/17a5a858e2045c6689a7e2f0cc4b77b90cdcf6f7